File: leetcode/getIntersectionNode.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function (headA, headB) {
  let setA = new Set();
  let curA = headA;
  while (curA) {
    setA.add(curA);
    curA = curA.next;
  }
  let curB = headB;
  while (curB) {
    if (setA.has(curB)) return curB;
    curB = curB.next;
  }
  return null;
};
var getIntersectionNode = function (headA, headB) {
  if (!headA || !headB) return null;
  let a = headA;
  let b = headB;

  // while (curA !== curB) {
  //   curA = curA.next;
  //   curB = curB.next;
  //   if (curA === curB) break;
  //   if (!curA) {
  //     curA = headB;
  //   }
  //   if (!curB) {
  //     curB = headA;
  //   }
  // }
  while (a !== b) {
    a = a === null ? headB : a.next;
    b = b === null ? headA : b.next;
  }

  return curA;
};

File: leetcode/getIntersectionNode.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        return intersection_node_2_pointer(headA, headB)


def intersection_node_2_pointer(headA, headB):
    c1 = headA
    c2 = headB
    while c1 != c2:
        c1 = headB if c1 == None else c1.next
        c2 = headA if c2 == None else c2.next
    return c1


def intersection_node_set():
    return
